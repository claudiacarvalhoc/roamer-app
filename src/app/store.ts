import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import appReducer from '../slice/appSlice';
import menuReducer from '../slice/menuSlice';

export const store = configureStore({
    reducer: {
        app: appReducer,
        menu: menuReducer,
    },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
    ReturnType,
    RootState,
    unknown,
    Action<string>
>;
