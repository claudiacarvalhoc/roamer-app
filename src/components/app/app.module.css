.app {
    margin-left: 10%;
  }

.project {
  margin-bottom: 40px;
}

.collapsed {
  display: flex;
  justify-content: flex-start;
}