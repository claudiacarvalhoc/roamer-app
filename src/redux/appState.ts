import { DefaultRootState } from "react-redux";

export interface AppState extends DefaultRootState {
    expand: boolean;
    projects: ProjectProperties[];
    menuTexts: MenuTexts,
    projectTexts: ProjectTextsProperties,
}

export interface MenuTexts {
    projectText: string;
    expandText: string;
    collapseText: string;
}

export interface ProjectTextsProperties {
    doneText: string;
    baseWordsText: string;
    teamText: string;
    keysText: string;
    qaIssuesText: string;
}

export interface IconType {
    Icon: 'up' | 'down' | 'check' | 'user' | 'rate' | 'camera' | 'book';
}

export interface TranslationSectionsProperties {
    country: string;
    language: string;
    done: number;
    wordsToDo: number;
    unverified: number;
}

export interface ProjectProperties {
    id: number;
    name: string;
    done: number;
    baseWords: number;
    team: number;
    keys: number;
    issues: number;
    icons: string[];
    badges: string[];
    translationSections: TranslationSectionsProperties[];
}
