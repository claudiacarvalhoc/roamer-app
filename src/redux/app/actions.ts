import { Action } from 'redux';
import { ProjectProperties } from "../appState";
import { ADD_PROJECT, EXPAND_PROJECTS, COLLAPSE_PROJECTS } from "../types";

export interface addProjectAction extends Action<string> {
    project: ProjectProperties;
}

export interface ExpandProjectsAction extends Action {
}

export interface CollapseProjectAction extends Action {

}

export const addProjectAction = (
    project: ProjectProperties
): addProjectAction => ({
    type: ADD_PROJECT,
    project,
});

export const expandProjectsAction = (): ExpandProjectsAction => ({
    type: EXPAND_PROJECTS,
});

export const collapseProjectAction = (): CollapseProjectAction => ({
    type: COLLAPSE_PROJECTS,
});

export type AppActions =
    | addProjectAction;