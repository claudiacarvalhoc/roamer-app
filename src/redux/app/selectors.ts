import { MenuTextsState, ProjectState, ProjectTextsState } from "../appState";
import { RootState } from "../reducers";

export const isExpanded = (state: RootState): boolean => state.app.expand;
export const projects = (state: RootState): ProjectState[] => state.app.projects;
export const menuTexts = (state: RootState): MenuTextsState => state.app.MenuTextsState;
export const projectTexts = (state: RootState): ProjectTextsState => state.app.projectTexts;