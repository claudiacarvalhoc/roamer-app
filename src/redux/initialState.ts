import { AppState, MenuTextsState, ProjectState, ProjectTextsState } from "./appState";

export const menuInitialState: MenuTextsState = {
    projectText: 'New project ⇧⌘P',
    expandText: 'Expand all',
    collapseText: 'Collapse all',
};

export const projectsInitialState: ProjectTextsState = {
    doneText: 'DONE',
    baseWordsText: 'BASE WORDS',
    teamText: 'TEAM',
    keysText: 'KEYS',
    qaIssuesText: 'QA ISSUES',
    wordsToDoText: 'WORDS TO DO',
    unverifiedText: 'UNVERIFIED',
    addLanguageText: 'Add language',
};

export const projectInitialState: ProjectState = {
    id: 1,
    name: 'Roamer App (Android SDK test)',
    done: 30,
    baseWords: 160246,
    team: 26,
    keys: 1834,
    issues: 47514,
    icons: ['up', 'down', 'check', 'user', 'rate', 'camera', 'book'],
    badges: [
        {
            name: 'Roamer',
            type: 'orange',
        },
        {
            name: 'iOS',
            type: 'green',
        }
    ],
    translationSections: [{
        id: 1,
        country: 'uk',
        language: 'English',
        done: 20,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 2,
        country: 'uk',
        language: 'English',
        done: 20,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 3,
        country: 'uk',
        language: 'English',
        done: 20,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 4,
        country: 'uk',
        language: 'English',
        done: 20,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 5,
        country: 'ru',
        language: 'Russian',
        done: 45,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 6,
        country: 'ru',
        language: 'Russian',
        done: 45,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 7,
        country: 'ru',
        language: 'Russian',
        done: 45,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 8,
        country: 'ru',
        language: 'Russian',
        done: 45,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 9,
        country: 'it',
        language: 'Italian',
        done: 83,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 10,
        country: 'it',
        language: 'Italian',
        done: 83,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 11,
        country: 'it',
        language: 'Italian',
        done: 83,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 12,
        country: 'it',
        language: 'Italian',
        done: 83,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 13,
        country: 'mk',
        language: 'Macedonian',
        done: 67,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 14,
        country: 'mk',
        language: 'Macedonian',
        done: 67,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 15,
        country: 'mk',
        language: 'Macedonian',
        done: 67,
        wordsToDo: 16246,
        unverified: 50000,
    }]
};

export const appInitialState: AppState = {
    expand: true,
    projects: [
        projectInitialState
    ],
    MenuTextsState: menuInitialState,
    projectTexts: projectsInitialState,
};

export const newProjectInitialState: ProjectState = {
    id: 0,
    name: 'Roamer App (Android SDK test)',
    done: 0,
    baseWords: 160246,
    team: 26,
    keys: 1834,
    issues: 47514,
    icons: ['up', 'down', 'check', 'user', 'rate', 'camera', 'book'],
    badges: [
        {
            name: 'Roamer',
            type: 'orange',
        },
        {
            name: 'iOS',
            type: 'green',
        }
    ],
    translationSections: [],
};