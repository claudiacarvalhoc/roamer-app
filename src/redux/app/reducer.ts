// import { ActionTypes } from '../types';
import { Reducer, combineReducers } from 'redux';
import { AppActions } from './actions';
import { PROJECT_LOAD_SUCCESS } from "../types";
import { AppState } from '../appState';

const appInitialState: AppState = {
    expand: true,
    projects: [],
};

export const appReducer = (
    state: AppState = appInitialState,
    action: AppActions
): AppState => {
    switch (action.type) {
        case PROJECT_LOAD_SUCCESS:
            return {
                ...state,
                projects: action.projects,
            };
        default:
            return {
                ...state,
            };
    }
};
