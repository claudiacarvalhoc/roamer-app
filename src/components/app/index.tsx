import React, { FC, useEffect } from 'react';
import { connect } from 'react-redux';
import { ProjectProperties } from '../../redux/appState';
import './styles.css';
import Menu from '../menu';
import Project from '../project';
import { fecthProjects } from '../../redux/fetch';
import { AppDispatch } from '../../redux/store';
import { RootState } from '../../redux/reducers';
import { isExpanded, projects } from '../../redux/app/selectors';

export interface AppOwnProps {

}

export interface AppStateProps {
  expand: boolean;
  projects: ProjectProperties[];
}

export interface AppDispatchProps {
  fetchProjects: () => void;
}

export type AppProps = AppOwnProps & AppStateProps & AppDispatchProps;

const App: FC<AppProps> = ({ fetchProjects }) => {
  useEffect(() => {
    fetchProjects();
  }, []);


  return (
    <div className="App">
      <Menu />
      <Project />
    </div>
  );
};

const mapStateToProps = (state: RootState): AppStateProps => ({
  expand: isExpanded(state),
  projects: projects(state),
});

const mapDispatchToProps = (
  dispatch: AppDispatch
): AppDispatchProps => ({
  fetchProjects: () => dispatch(fecthProjects())
});

export default connect<AppStateProps, AppDispatchProps>(
  mapStateToProps,
  mapDispatchToProps
)(App);