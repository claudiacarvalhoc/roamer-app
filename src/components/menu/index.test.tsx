import React from 'react';
import Menu from './index';
import Enzyme, { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import { appInitialState } from '../../redux/initialState';
import { Provider } from 'react-redux';
import { expandProjectsAction, collapseProjectAction } from '../../redux/app/actions';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
// import { createProject } from '../../redux/fetch';

import thunk from 'redux-thunk'

Enzyme.configure({ adapter: new Adapter() });

jest.mock('../../redux/app/actions');
jest.mock('../../redux/fetch');

const mockAction = () => ({ type: 'MOCK' });
const mockFetch = () => ({});

const getMountWrapper = () => {
    const store = configureMockStore([thunk])({
        app: appInitialState
    });
    return mount(
        <Provider store={store}>
            <Menu />
        </Provider>
    );
}

describe('<Menu />', () => {
    it('render correctly', () => {
        const component = getMountWrapper();
        expect(toJson(component)).toMatchSnapshot();
    });
    it('should dispatch the expand action when clicking on expand button', () => {
        (expandProjectsAction as jest.Mock).mockImplementationOnce(mockAction);
        const component = getMountWrapper();
        component
            .find('[data-test="button-expand"]')
            .first()
            .simulate('click');
        expect(expandProjectsAction).toHaveBeenCalledTimes(1);
    });
    it('should dispatch the collapse action when clicking on collapse button', () => {
        (collapseProjectAction as jest.Mock).mockImplementationOnce(mockAction);
        const component = getMountWrapper();
        component
            .find('[data-test="button-collapse"]')
            .first()
            .simulate('click');
        expect(collapseProjectAction).toHaveBeenCalledTimes(1);
    });
    // Warning the idea here was to test that an async function was invoked once
    // But, even registring the redux-trunk on store the error is the same
    // Actions must be plain objects. Use custom middleware for async actions.
    // it('should dispatch the new project action when clicking on new project button', () => {
    //     (createProject as jest.Mock).mockImplementationOnce(mockFetch);
    //     const component = getMountWrapper();
    //     component
    //         .find('[data-test="button-collapse"]')
    //         .first()
    //         .simulate('click');
    //     expect(createProject).toHaveBeenCalledTimes(1);
    // });
});
