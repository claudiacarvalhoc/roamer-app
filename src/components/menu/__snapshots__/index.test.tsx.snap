// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> render correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Menu)>
    <Menu
      addProject={[Function]}
      collapse={[Function]}
      collapseText="Collapse all"
      expand={[Function]}
      expandText="Expand all"
      projectText="New project ⇧⌘P"
    >
      <div
        className="menu"
      >
        <Button
          data-test="button-new-project"
          onClick={[Function]}
          text="New project ⇧⌘P"
          type="primary"
        >
          <button
            className="button primary"
            onClick={[Function]}
          >
            New project ⇧⌘P
          </button>
        </Button>
        <Button
          data-test="button-expand"
          onClick={[Function]}
          text="Expand all"
          type="secondary"
        >
          <button
            className="button secondary"
            onClick={[Function]}
          >
            Expand all
          </button>
        </Button>
        <Button
          data-test="button-collapse"
          onClick={[Function]}
          text="Collapse all"
          type="secondary"
        >
          <button
            className="button secondary"
            onClick={[Function]}
          >
            Collapse all
          </button>
        </Button>
      </div>
    </Menu>
  </Connect(Menu)>
</Provider>
`;
